File: forms/fall-risk-assessment/back-end-with-loco/src/engine/fall_risk_grader.rs
//! Fall risk grader module.

// Fall Risk grader. Pure functions: take an `AssessmentData`, return the
// total Morse Fall Scale score (0..=125), the `Severity`, the list of
// fired item rules, and the Critical-override metadata.
//
// Severity cutoffs (raw MFS):
//   0..=24   -> low
//   25..=44  -> moderate
//   >=45     -> high
//
// Critical-override (raises the severity to "critical" regardless of raw
// MFS):
//   - recurrent falls with injury
//   - anticoagulated patient
//   - MFS >= 75

use super::flagged_issues::detect_additional_flags;
use super::mfs_rules::{
    MfsItem, has_recurrent_falls_with_injury, is_anticoagulated, mfs_items,
};
use super::types::{AssessmentData, FiredRule, GradingResult, Severity};

/// Classify a numeric MFS score (0..=125) into the base severity. The
/// Critical-override is applied separately by [`grade`].
pub fn classify_mfs_score(score: i32) -> Severity {
    if score >= 45 {
        "high".to_string()
    } else if score >= 25 {
        "moderate".to_string()
    } else {
        "low".to_string()
    }
}

/// Look up the option label corresponding to an MFS item's selected
/// score. Returns the empty string if no option matches.
fn mfs_option_label(item: &MfsItem, score: i32) -> &'static str {
    for opt in item.options {
        if opt.score == score {
            return opt.label;
        }
    }
    ""
}

/// Evaluate the six-item Morse Fall Scale against the supplied
/// assessment data and apply the Critical-override. Mirrors
/// `gradeFallRisk` in the JS front-end engine.
pub fn grade_fall_risk(data: &AssessmentData) -> (i32, Severity, bool, Vec<String>, u32, Vec<FiredRule>) {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut mfs_score: i32 = 0;
    let mut answered_count: u32 = 0;

    for item in mfs_items() {
        if let Some(value) = (item.field_get)(&data.mfs) {
            answered_count += 1;
            mfs_score += value;
            let label = mfs_option_label(&item, value);
            let description = if label.is_empty() {
                item.description.to_string()
            } else {
                label.to_string()
            };
            fired_rules.push(FiredRule {
                id: item.id.to_string(),
                category: item.label.to_string(),
                description,
                score: value,
            });
        }
    }

    let base_severity = classify_mfs_score(mfs_score);

    let mut critical_reasons: Vec<String> = Vec::new();
    if has_recurrent_falls_with_injury(data) {
        critical_reasons.push("Recurrent falls with injury".to_string());
    }
    if is_anticoagulated(data) {
        critical_reasons.push("Anticoagulated patient".to_string());
    }
    if mfs_score >= 75 {
        critical_reasons.push(format!("MFS {mfs_score} (>= 75)"));
    }

    let critical_override = !critical_reasons.is_empty();
    let severity = if critical_override {
        "critical".to_string()
    } else {
        base_severity
    };

    (mfs_score, severity, critical_override, critical_reasons, answered_count, fired_rules)
}

/// Run the grader and bundle the flag-detection output into a
/// [`GradingResult`] with an ISO-8601 timestamp. This is the public
/// entry point used by the controllers.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let (mfs_score, severity, critical_override, critical_reasons, answered_count, fired_rules) =
        grade_fall_risk(data);
    let additional_flags = detect_additional_flags(data);
    GradingResult {
        mfs_score,
        severity,
        critical_override,
        critical_reasons,
        answered_count,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/fall-risk-assessment/front-end-with-svelte/src/lib/engine/fall-risk-grader.ts
// Fall Risk grader. Pure functions: take an `AssessmentData` object and return
// the total Morse Fall Scale score (0-125), the `Severity`, the list of fired
// item rules, the Critical-override metadata, and the additional flags.
//
// Severity cutoffs (raw MFS):
//   0-24   -> low
//   25-44  -> moderate
//   >=45   -> high
//
// Critical-override (raises severity to 'critical' regardless of raw MFS):
//   - recurrent falls with injury
//   - anticoagulated patient
//   - MFS >= 75

import type { AssessmentData, FiredRule, GradingResult, MfsItem, Severity } from './types';
import { mfsItems, ancillaryRules } from './mfs-rules';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Classify a numeric MFS score (0-125) into the base severity. The
 * Critical-override is applied separately by `calculateFallRiskGrade`.
 */
export function classifyMfsScore(score: number): Severity {
	if (score >= 45) return 'high';
	if (score >= 25) return 'moderate';
	return 'low';
}

/** Look up the option label for an MFS field's selected score. */
function mfsOptionLabel(item: MfsItem, score: number | null): string {
	if (score === null || score === undefined) return '';
	const opt = item.options.find((o) => o.score === score);
	return opt ? opt.label : '';
}

/**
 * Evaluate the six-item Morse Fall Scale against the supplied assessment data,
 * apply the Critical-override, and collect the additional flags.
 */
export function calculateFallRiskGrade(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	let mfsScore = 0;
	let answeredCount = 0;

	for (const item of mfsItems) {
		const value = data.mfs ? data.mfs[item.field] : null;
		if (value !== null && value !== undefined) {
			answeredCount++;
			mfsScore += value;
			firedRules.push({
				id: item.id,
				category: item.label,
				description: mfsOptionLabel(item, value) || item.description,
				score: value
			});
		}
	}

	const baseSeverity = classifyMfsScore(mfsScore);

	const criticalReasons: string[] = [];
	if (ancillaryRules.hasRecurrentFallsWithInjury(data)) {
		criticalReasons.push('Recurrent falls with injury');
	}
	if (ancillaryRules.isAnticoagulated(data)) {
		criticalReasons.push('Anticoagulated patient');
	}
	if (mfsScore >= 75) {
		criticalReasons.push(`MFS ${mfsScore} (>= 75)`);
	}

	const criticalOverride = criticalReasons.length > 0;
	const severity: Severity = criticalOverride ? 'critical' : baseSeverity;

	return {
		mfsScore,
		severity,
		criticalOverride,
		criticalReasons,
		answeredCount,
		firedRules,
		additionalFlags: detectAdditionalFlags(data),
		timestamp: new Date().toISOString()
	};
}
